### PR TITLE
Open local links based on their MIME type

### PR DESCRIPTION
**Pull request by fnevgeny**

### CODE REVIEW

File: src/classes/action/link_action.vala [hunk 1]
```diff
@@ -114,6 +114,33 @@ namespace pdfpc {
             );
         }
 
+        /**
+         * Launch an application, trying to cover all corner cases.
+         */
+        private void launch_for_uri(string uri) throws GLib.Error {
+            if (Uri.parse_scheme(uri) != null) {
+                AppInfo.launch_default_for_uri(uri, null);
```
Comment (AndreasBilke) [inline src/classes/action/link_action.vala:122]: I just read over the docs and `launch_default_for_uri_async` looks more promising to me?

File: src/classes/action/link_action.vala [hunk 2]
```diff
@@ -114,6 +114,33 @@ namespace pdfpc {
...
+                var appinfo = AppInfo.get_default_for_type(ctype, false);
+                string path;
+                if (Path.is_absolute(uri)) {
+                    path = uri;
+                } else {
+                    // If the path is not absolute, translate it relative
+                    // to the PDF document location
+                    var pdf_fname = controller.get_pdf_fname();
+                    var dirname = Path.get_dirname(pdf_fname);
+                    path = Posix.realpath(Path.build_filename(dirname, uri));
```
Comment (AndreasBilke) [inline src/classes/action/link_action.vala:136]: We do not officially support Windows, but is `Posix` working there? Can we use `Glib.File.new_build_filename()` instead?

File: src/classes/action/link_action.vala (hunk 1 above)
Comment (fnevgeny) [reply src/classes/action/link_action.vala:122]: Do you think it's reasonable for pdfpc to provide any type of GUI selector to the user during a presentation?

More generally, I don't quite understand the need for clickable URI links; it's against the pdfpc principles to show identical views on two screens...

Comment (AndreasBilke) [reply src/classes/action/link_action.vala:122]: You never know what weird setups people have. And as far as I understand the "select application GUI" only shows up if you have no default.

I use sometimes clickable URI (mostly web pages, sometimes other files on my disk) in my presentations.

Comment (fnevgeny) [reply src/classes/action/link_action.vala:122]: Well, but then the web page appears on your screen, not where the people in the auditorium can see it?

**Comment by fnevgeny** [comment]
So far, I couldn't make `launch_default_for_uri_async()` work. That's, when a link is recognized, it works. But nothing happens in the case of failure (no signal fired). The sync variant at least throws an `Error`...

**Review by AndreasBilke — APPROVED** [review]

**Comment by fnevgeny** [comment]
OK, I managed to use `launch_default_for_uri_async()` (previously, invoked it incorrectly). But still, the result is exactly the same - just the "Operation not supported" error thrown.

**Pull request merged**